jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp

```diff
@@ -177,6 +191,52 @@ ecma_builtin_json_parse_string (ecma_json_token_t *token_p) /**< token argument
     *write_p++ = *current_p++;
   }
 
+  /*
+   * Post processing surrogate pairs.
+   *
+   * The general issue is, that surrogate fragments can come from
+   * the original stream and can be constructed by \u sequences
+   * as well. We need to construct code points from them.
+   *
+   * Example: JSON.parse ('"\\ud801\udc00"') === "\ud801\udc00"
+   *          The first \u is parsed by JSON, the second is by the lexer.
+   *
+   * The rewrite happens in-place, since the write pointer is always
+   * precede the read-pointer. We also cannot create an UTF8 iterator,
+   * because the lit_is_utf8_string_valid assertion may fail.
+   */
+
+  lit_utf8_byte_t *read_p = token_p->u.string.start_p;
+  lit_utf8_byte_t *read_end_p = write_p;
+  write_p = read_p;
+
+  while (read_p < read_end_p)
+  {
+    if (read_p[0] == LIT_UTF8_SURROGATE_START_BYTE
+        && (read_end_p - read_p) >= 6
+        && read_p[3] == LIT_UTF8_SURROGATE_START_BYTE)
+    {
+      lit_code_point_t first_code_point, second_code_point;
+
+      lit_read_code_point_from_utf8 (read_p, 3, &first_code_point);
+      JERRY_ASSERT (first_code_point >= 0xd000 && first_code_point <= 0xdfff);
```

Comment:
I expected that this will raise a lot of questions. I was thinking about this yesterday evening, and thinking about another way of doing this. I will submit a patch soon.